pkg/pillar/pubsub/socketdriver/driver.go

```diff
@@ -218,3 +228,54 @@ func (s *SocketDriver) fixedDirName(name string) string {
 func (s *SocketDriver) persistentDirName(name string) string {
 	return fmt.Sprintf("%s/status/%s", "/persist", name)
 }
+
+//getBuffer returns a buffer and a done func to call at defer.
+func getBuffer() (*[]byte, func()) {
+	buf := bufPool.Get().(*[]byte)
+	return buf, func() {
+		bufPool.Put(buf)
+	}
+}
+
+// check and waits till conn's fd is readable
+func connReadCheck(conn net.Conn) error {
+	var sysErr error
+
+	sysConn, ok := conn.(syscall.Conn)
+	if !ok {
+		log.Error("connReadCheck: Not syscall.Conn")
+		return nil
+	}
+	rawConn, err := sysConn.SyscallConn()
+	if err != nil {
+		err := fmt.Errorf("connReadCheck: Exception while getting rawConn: %s", err.Error())
+		log.Error(err)
+		return err
+	}
+
+	err = rawConn.Read(func(fd uintptr) bool {
+		_, err := syscall.Read(int(fd), []byte{})
```

Comment:
Note that this would block the thread in the kernel and we don't know what impact that will have when we have one hundred or more go routines doing the same thing. We want to have the go runtime to the blocking without consuming threads.

Reply:
Actually I was wrong. The rawConn.Read does not block in syscall.Read - the underlying fd is non-blocking hence it gets to the EAGAIN case and it waits/polls in the go runtime (observed by pkill -USR1 zedbox). Thus this approach should be fine.

However, you need to call syscall.Recvfrom instead of syscall.Read as in
+               _, _, err := syscall.Recvfrom(int(fd), []byte{}, 0)

With that I see that the connReadCheck function doesn't return until there is data to read i.e., the Read in serveConnection etc does not block/sleep waiting for data.

Reply:
I've tried various implementations (including one which reads and remembers the first byte before allocating the buffer and doesn't even use sync.Pool) and they all my qemu device extremely slow. I don't know if that is due to low memory or if something else is going on. @adarsh-zededa if you have suggestions we should chat.